torch_xla/csrc/aten_xla_type.cpp

```diff
@@ -1085,6 +1084,19 @@ at::Tensor XLANativeFunctions::dot(const at::Tensor& self,
       bridge::GetXlaTensor(self), bridge::GetXlaTensor(tensor)));
 }
 
+at::Tensor XLANativeFunctions::einsum(c10::string_view equation,
+                                      at::TensorList tensors) {
+  XLA_FN_COUNTER("xla::");
+  // Einsum operations with more than 2 operands, like bilinear operations, are
+  // not currently supported in XLA
+  if (tensors.size() > 2) {
```

Comment:
@bdhirsh This is a bit tricky, we want to overwrite `einsum` but we can only support a certain type of einsum. Looking at https://github.com/pytorch/xla/blob/master/torch_xla/csrc/aten_autograd_ops.cpp#L31 it seems like we actually need to let pt/xla to take care falling back for both forward and backward. However there is no such thing as `einsum_backward`, what are we going to do with backward case?

Reply:
on a second thought we should not fallback but call `at::native` function like https://github.com/pytorch/xla/blob/master/torch_xla/csrc/aten_xla_type.cpp#L3086 which will redecompose einsum into smaller ops. I think this should solve the backward issue for unsupported einsum too.

Reply:
Yes, I think you want to fall back by calling into `at::native::einsum`. Falling back to CPU seems like a big pessimisation; instead, you just want to fall back to the existing decomposition in core, and run XLA on the decomposed ops.

Reply:
I also... think that should fix the `requires_grad` issue that you're seeing, but it's worth confirming 😛.

Reply:
Updated to `at::native::einsum` here, but there's still the issue of `requires_grad` for cases with 1 or 2 operands. We want to use `aten_autograd_ops::EinsumAutogradFunction::apply` there to leverage XLA's einsum implementation.

Reply:
ah, so it's working in the case where you call `at::native::einsum`, but it's losing the requires_grad in the case where it calls the custom autograd function?

If it's an issue with the autograd function, what I would check first is:
(1) Did you implement the autograd function the same way that the existing ones in pytorch_xla are implemented? (e.g. `max_pool2d`?
(2) If so, can you repro the issue `max_pool2d` too?

Reply:
As far as I know, this einsum autograd function is implemented the same way as the max_pool functions. However, I cannot reproduce the issue with max_pool2d.

Max pool 2d was overriden as an autograd function here https://github.com/pytorch/xla/pull/2236. Since then it looks like we've removed `scripts/gen.py`, but I believe all I have to do to setup the code generation now is add `einsum` to the `autograd` section of `xla_native_functions.yaml`. Could there be additional setup required, either in this repo or the parent repo? CC @JackCaoG

Reply:
It doesn't look like we explicitly do anything in the max_pool implementations to forward `requires_grad` to the output. Where is that happening?

Reply:
@bdhirsh I think one difference is that pytorch does not have backward formula for einsum, it will always get dispatched to smaller ops. Not sure if this will affect the behavior of `require_grads`.